### PR TITLE
Add git commit display to builds table

### LINKED ISSUE #2948: Display git commit ID on list of builds
build.snapcraft.io currently lists a git commit ID alongside build numbers so you can see which commit triggered the build. The new UI on snapcraft.io does not and I couldn't find an obvious way to figure out which commit triggered a build, unless I've missed something. This makes tracking down regressions much harder.

It would be great to have this feature in the new UI, if not already planned.

### PR DESCRIPTION
**Pull request by gajeshbhat**
## Done

- Added git commit ID display to builds table
- Created utility functions for commit formatting and GitHub URL generation  
- Added clickable links to GitHub commits when repository is connected
- Updated both individual build pages and builds list view
- Added comprehensive test coverage for all new functionality
- Handles edge cases (no commit ID, no GitHub repo, invalid data)

## How to QA

1. **Set up a snap with GitHub integration:**
   - Create a snap connected to a GitHub repository via Launchpad
   - Trigger builds by pushing commits to the repository

2. **Test builds table:**
   - Navigate to `/your-snap-name/builds`
   - Verify "Git commit" column appears in the builds table
   - Verify commit IDs are displayed as 7-character truncated hashes
   - Verify commit IDs are clickable links (if GitHub repo connected)
   - Verify links open correct GitHub commit pages in new tab

3. **Test individual build page:**
   - Click on any build in the builds table
   - Verify "Git commit" row appears in build details
   - Verify same formatting and linking behavior

4. **Test edge cases:**
   - Builds without commit IDs should show empty cells
   - Builds without GitHub repository should show plain text (no links)

## Testing

- [x] This PR has tests
  - Added comprehensive unit tests for `formatCommitId()` function (16 test cases)
  - Added comprehensive unit tests for `createGitHubCommitUrl()` function (6 test cases)  
  - Added comprehensive unit tests for `GitCommitLink` React component (10 test cases)
  - Added integration tests for backend data flow in `test_builds.py`
  - All tests pass in CI pipeline

## Issue / Card

Fixes #2948 

## Screenshots
I need some help with testing this. Could not find any helpful documentation about setting up Launchpad to help test this. I was able to set up my own GitHub App and pass the OAuth-based login. That's as far as I could get. Can you folks please help ?

### CODE REVIEW

File: webapp/publisher/snaps/build_views.py [hunk 1]
```diff
@@ -32,6 +32,12 @@ def get_builds(lp_snap, selection):
     snap_builds = []
     builders_status = None
 
+    # Extract GitHub repository info for commit links
+    github_repository = None
+    if lp_snap.get("git_repository_url"):
+        # Remove "https://github.com/" prefix to get owner/repo
+        github_repository = lp_snap["git_repository_url"][19:]
```
Comment (edisile) [inline webapp/publisher/snaps/build_views.py]: I get where the `[19:]` comes from, but it looks somewhat brittle... A regex match along these lines might be more resilient

```suggestion
        match = re.search("github.com/(?P<repo>.+/.+)", lp_snap["git_repository_url"])
        if match:
            github_repository = match.groupdict()["repo"]
```

File: webapp/publisher/snaps/build_views.py [hunk 2]
```diff
@@ -117,6 +124,13 @@ def get_snap_build(snap_name, build_id):
     lp_build = launchpad.get_snap_build(details["snap_name"], build_id)
 
     if lp_build:
+        # Get snap info to extract GitHub repository
+        lp_snap = launchpad.get_snap_by_store_name(details["snap_name"])
+        github_repository = None
+        if lp_snap and lp_snap.get("git_repository_url"):
+            # Remove "https://github.com/" prefix to get owner/repo
+            github_repository = lp_snap["git_repository_url"][19:]
```
Comment (edisile) [inline webapp/publisher/snaps/build_views.py]: Same as above, maybe move this logic to a helper function

File: static/js/publisher/utils/formatGitCommit.tsx [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+export function GitCommitLink({
+  commitId,
+  githubRepository,
+  showLink = true,
+}: GitCommitLinkProps): React.JSX.Element | null {
+  const formattedCommitId = formatCommitId(commitId);
+
+  if (!formattedCommitId) {
+    return null;
```
Comment (edisile) [inline static/js/publisher/utils/formatGitCommit.tsx]: Returning null isn't that great IMO, it means leaving the grid cell empty and this might confuse users. Returning a <span> like the other cases would be better, even more if you added something like an `aria-label` to help screen reader users get some context on why this cell isn't filled

File: static/js/publisher/utils/formatGitCommit.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+    <a
+      href={commitUrl}
+      target="_blank"
+      rel="noopener noreferrer"
+      className="p-text--code"
+      title={`View commit ${formattedCommitId} on GitHub`}
+    >
+      {formattedCommitId}
+    </a>
+  );
```
Comment (edisile) [inline static/js/publisher/utils/formatGitCommit.tsx]: A bit of a nitpick: these multiple return statements are unnecessary IMO, handling them separately doesn't really add much value so you could merge them into a single return at the end like

```tsx
  return (
    (!formattedCommitId) ? (
       <span
         className="p-text--code"
         aria-label="No commit data available for this build"
       >
         -
       </span>
    )
    (!showLink || !githubRepository || !commitUrl) ? (
       <span className="p-text--code">{formattedCommitId}</span>
    ) : (
      <a
        href={commitUrl}
        target="_blank"
        rel="noopener noreferrer"
        className="p-text--code"
        title={`View commit ${formattedCommitId} on GitHub`}
        aria-label={`View commit ${formattedCommitId} on GitHub`}
      >
        {formattedCommitId}
      </a>
    )
  );
```

File: static/js/publisher/utils/__tests__/formatGitCommit.test.tsx
Comment (edisile) [inline static/js/publisher/utils/__tests__/formatGitCommit.test.tsx:1]: Nice coverage!

File: static/js/publisher/utils/__tests__/formatGitCommit.test.tsx [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+  test("renders nothing when commit ID is undefined", () => {
+    const { container } = render(<GitCommitLink commitId={undefined} />);
+    expect(container.firstChild).toBeNull();
+  });
+
+  test("renders nothing when commit ID is too short", () => {
+    const { container } = render(<GitCommitLink commitId="abc123" />);
+    expect(container.firstChild).toBeNull();
+  });
```
Comment (edisile) [inline static/js/publisher/utils/__tests__/formatGitCommit.test.tsx]: It might not be a good idea to render `null` in these cases, check comments below for more info

**Comment by edisile** [comment]
The `cla-check` action failed so I guess this is also your first contribution to a Canonical repository, so congrats again!

To get this PR merge we will need you to accept the [Contributor License Agreement](https://ubuntu.com/legal/contributors). No need to do it immediately but it might be worth doing it now rather than later just to speed things up and merge this quicker when it's ready

File: webapp/publisher/snaps/build_views.py (hunk 1 above)
Comment (gajeshbhat) [reply webapp/publisher/snaps/build_views.py]: Fixed! Replaced hardcoded string slicing with regex parsing using `extract_github_repository()` helper function. Now handles .git suffixes, trailing slashes, and invalid URLs properly.

File: webapp/publisher/snaps/build_views.py (hunk 2 above)
Comment (gajeshbhat) [reply webapp/publisher/snaps/build_views.py]: Fixed. Details in the comment above.

File: static/js/publisher/utils/formatGitCommit.tsx (hunk 3 above)
Comment (gajeshbhat) [reply static/js/publisher/utils/formatGitCommit.tsx]: Fixed! Now returns a dash (-) with `aria-label="No commit data available for this build"` instead of null.

File: static/js/publisher/utils/formatGitCommit.tsx (hunk 4 above)
Comment (gajeshbhat) [reply static/js/publisher/utils/formatGitCommit.tsx]: Consolidated into a single return with ternary operators as suggested. Also added aria-label to the link.

File: static/js/publisher/utils/__tests__/formatGitCommit.test.tsx
Comment (gajeshbhat) [reply static/js/publisher/utils/__tests__/formatGitCommit.test.tsx:1]: Thank you.

File: static/js/publisher/utils/__tests__/formatGitCommit.test.tsx (hunk 5 above)
Comment (gajeshbhat) [reply static/js/publisher/utils/__tests__/formatGitCommit.test.tsx]: Updated all tests to expect dash (-) with proper aria-label instead of null. Also took some time to add test coverage for the new helper function.

**Comment by gajeshbhat** [comment]
> The `cla-check` action failed so I guess this is also your first contribution to a Canonical repository, so congrats again!
> 
> To get this PR merge we will need you to accept the [Contributor License Agreement](https://ubuntu.com/legal/contributors). No need to do it immediately but it might be worth doing it now rather than later just to speed things up and merge this quicker when it's ready

Thank you @edisile for the feedback and for taking the time to review my PR. I have signed the CLA by submitting the form using the link you provided. Please let me know if you have further questions.

**Review by edisile — CHANGES_REQUESTED** [review]
LGTM 👍
Just one last thing, the JS link check is failing because of a formatting issue. Could you please run `npx eslint static/js --fix` and push again? Thanks

**Comment by gajeshbhat** [comment]
> LGTM 👍 Just one last thing, the JS link check is failing because of a formatting issue. Could you please run `npx eslint static/js --fix` and push again? Thanks

Thank you for pointing it out. Fixed the linting issue and pushed the changes. Thanks again for the review :)

**Comment by gajeshbhat** [comment]
Thank you for the approval @edisile ! Looking forward to contributing more to the Snap ecosystem. Cheers.
